File: backend/util/video_limits.py
"""视频发布校验规则（单点数据源）"""

import math


# 单位：秒 / bytes
# 数字必须与 docs/superpowers/specs/2026-06-19-video-validation-and-required-fields-design.md 第 2 节一致
VIDEO_LIMITS: dict[str, dict] = {
    "tencent_video": {"min_duration": 5,    "max_duration": 5400,             "max_size": 20 * 1024**3, "max_title_length": 80},           # 5s~90min,  20G,标题≤80字
    "iqiyi":         {"min_duration": 5,    "max_duration": 3600,             "max_size": 16 * 1024**3, "max_title_length": math.inf},  # 5s~60min,  16G
    "douyin":        {"min_duration": 5,    "max_duration": 3600,             "max_size": 16 * 1024**3, "max_title_length": math.inf},  # 5s~60min,  16G
    "baijiahao":     {"min_duration": 5,    "max_duration": math.inf,         "max_size": 12 * 1024**3, "max_title_length": math.inf},  # 5s~无,   12G
    "weibo":         {"min_duration": 0,    "max_duration": math.inf,         "max_size": 15 * 1024**3, "max_title_length": 30},           # 无时长下限,  15G,标题≤30字
    "kuaishou":      {"min_duration": 5,    "max_duration": 3600,             "max_size": 12 * 1024**3, "max_title_length": math.inf},  # 5s~60min,  12G
    "bilibili":      {"min_duration": 5,    "max_duration": 36000,            "max_size": 16 * 1024**3, "max_title_length": 80, "max_desc_length": 2000},  # 5s~600min,16G,标题≤80字,简介≤2000字(emoji=3)
    "xiaohongshu":   {"min_duration": 5,    "max_duration": 14400,            "max_size": 20 * 1024**3, "max_title_length": 20},          # 5s~240min,20G,标题≤20字
    "channels":      {"min_duration": 5,    "max_duration": 28800,            "max_size": 20 * 1024**3, "max_title_length": math.inf},  # 5s~480min,20G
    "tiktok":        {"min_duration": 5,    "max_duration": 3600,             "max_size": 16 * 1024**3, "max_title_length": math.inf},  # 5s~60min,  16G
    "youtube":       {"min_duration": 5,    "max_duration": 36000,            "max_size": 16 * 1024**3, "max_title_length": math.inf},  # 5s~600min,16G
    "alipay":        {"min_duration": 5,    "max_duration": math.inf,         "max_size": 8 * 1024**3,  "max_title_length": math.inf},   # 5s~无,    8G(文档:≤8G,时长不限)
    "zhihu":         {"min_duration": 0,    "max_duration": math.inf,         "max_size": math.inf,     "max_title_length": math.inf},   # 文档:时长大小不限
    # CSDN: 视频大小≤2G, 时长不限, 标题≤30字
    "csdn":          {"min_duration": 0,    "max_duration": math.inf,         "max_size": 2 * 1024**3,  "max_title_length": 30},
    # VIVO: 视频大小≤2G, 时长≤90min(5400s), 用「视频描述」非标题故不限标题
    "vivo":          {"min_duration": 0, "max_duration": 5400,             "max_size": 2 * 1024**3,  "max_title_length": math.inf},
    # 微信公众号: 视频时长<1h, 标题≤64字, 描述(含#标签)≤300字
    "weixin_gzh":    {"min_duration": 0, "max_duration": 3600,             "max_size": math.inf,     "max_title_length": 64, "max_desc_length": 300},
    # 淘宝光合: 时长≤30min(1800s), 文件≤1.5G, 标题≤30字, 描述(含#标签)≤1000字
    "taobao_guanghe": {"min_duration": 0, "max_duration": 1800,            "max_size": int(1.5 * 1024**3), "max_title_length": 30, "max_desc_length": 1000},
    # 京东京麦: 标题 5~27 字
    "jingmai":        {"min_duration": 0, "max_duration": math.inf,        "max_size": math.inf,     "min_title_length": 5, "max_title_length": 27},
}


_PLATFORM_NAMES = {
    "tencent_video": "腾讯视频",
    "iqiyi": "爱奇艺",
    "douyin": "抖音",
    "baijiahao": "百家号",
    "weibo": "微博",
    "kuaishou": "快手",
    "bilibili": "B站",
    "xiaohongshu": "小红书",
    "channels": "视频号",
    "tiktok": "TikTok",
    "youtube": "YouTube",
    "alipay": "支付宝",
    "zhihu": "知乎",
    "csdn": "CSDN",
    "vivo": "VIVO",
    "weixin_gzh": "微信公众号",
    "taobao_guanghe": "淘宝光合",
    "jingmai": "京东京麦",
}


def _format_size(size_bytes: float) -> str:
    """自适应单位：B/KB/MB/GB。inf/nan/负数返回"未知"。"""
    if size_bytes is None or math.isnan(size_bytes) or math.isinf(size_bytes) or size_bytes < 0:
        return "未知"
    if size_bytes < 1024:
        return f"{size_bytes:.1f} B"
    if size_bytes < 1024**2:
        return f"{size_bytes / 1024:.1f} KB"
    if size_bytes < 1024**3:
        return f"{size_bytes / 1024**2:.1f} MB"
    return f"{size_bytes / 1024**3:.1f} GB"


def _format_duration(seconds: float) -> str:
    """自适应单位：秒 / 分秒 / 时分秒。inf/nan/负数返回"未知"。"""
    if seconds is None or math.isnan(seconds) or math.isinf(seconds) or seconds < 0:
        return "未知"
    seconds = int(seconds)
    if seconds < 60:
        return f"{seconds} 秒"
    h, rem = divmod(seconds, 3600)
    m, s = divmod(rem, 60)
    if h > 0:
        return f"{h} 小时 {m} 分 {s} 秒"
    return f"{m} 分 {s} 秒"


def _format_max_duration(max_duration: float) -> str:
    if max_duration == math.inf:
        return "无限制"
    return _format_duration(max_duration)


def validate_video_for_platform(platform_key: str, duration_sec: float, size_bytes: float) -> tuple[bool, str]:
    """校验视频时长和大小是否符合平台限制。

    Args:
        platform_key: 平台 key（如 "douyin"）
        duration_sec: 时长（秒）
        size_bytes: 大小（bytes）

    Returns:
        (ok, error_msg). error_msg 为空时表示通过。
        未配置的平台默认放行（新平台不阻塞）。
    """
    limits = VIDEO_LIMITS.get(platform_key)
    if limits is None:
        return True, ""

    name = _PLATFORM_NAMES.get(platform_key, platform_key)

    if duration_sec < limits["min_duration"]:
        return False, (
            f"{name}：时长 {_format_duration(duration_sec)} "
            f"小于最小值 ({_format_duration(limits['min_duration'])})"
        )
    if duration_sec > limits["max_duration"]:
        return False, (
            f"{name}：时长 {_format_duration(duration_sec)} "
            f"超出最大值 ({_format_max_duration(limits['max_duration'])})"
        )
    if size_bytes > limits["max_size"]:
        return False, (
            f"{name}：大小 {_format_size(size_bytes)} "
            f"超出限制 (最大 {_format_size(limits['max_size'])})"
        )
    return True, ""


def validate_title_for_platform(platform_key: str, title: str) -> tuple[bool, str]:
    """校验视频标题是否符合平台限制(如小红书 ≤ 20 字)。

    Args:
        platform_key: 平台 key
        title: 标题

    Returns:
        (ok, error_msg). error_msg 为空时表示通过。
        未配置的平台默认放行(无标题长度限制)。

    字符计算规则:BMP 字符 = 1,emoji 等非 BMP 字符 = 3。
    """
    limits = VIDEO_LIMITS.get(platform_key)
    if limits is None:
        return True, ""

    name = _PLATFORM_NAMES.get(platform_key, platform_key)
    min_len = limits.get("min_title_length", 0)
    max_len = limits.get("max_title_length", math.inf)

    # 按 emoji=3 规则计算字符数
    title_len = 0
    for ch in (title or ""):
        title_len += 3 if ord(ch) > 0xFFFF else 1

    if min_len and title_len < min_len:
        return False, f"{name}：标题至少 {min_len} 个字 (当前 {title_len} 字)"

    if max_len != math.inf and title_len > max_len:
        return False, (
            f"{name}：标题 {title_len} 字超过限制 (最多 {max_len} 字,emoji 按 3 算)"
        )
    return True, ""


def validate_desc_for_platform(platform_key: str, desc: str) -> tuple[bool, str]:
    """校验视频简介是否符合平台限制(如 B 站 ≤ 2000 字,emoji 按 3 算)。

    Args:
        platform_key: 平台 key
        desc: 简介内容

    Returns:
        (ok, error_msg). error_msg 为空时表示通过。
        未配置的平台默认放行。
    """
    limits = VIDEO_LIMITS.get(platform_key)
    if limits is None:
        return True, ""

    name = _PLATFORM_NAMES.get(platform_key, platform_key)
    max_len = limits.get("max_desc_length", math.inf)
    if max_len == math.inf:
        return True, ""

    # 按 emoji=3 规则计算字符数
    desc_len = 0
    for ch in (desc or ""):
        desc_len += 3 if ord(ch) > 0xFFFF else 1

    if desc_len > max_len:
        return False, (
            f"{name}：简介 {desc_len} 字超过限制 (最多 {max_len} 字,emoji 按 3 算)"
        )
    return True, ""

File: frontend/src/config/videoLimits.js
/**
 * 视频发布校验规则（前端镜像，与 backend/util/video_limits.py 保持一致）
 *
 * 修改时请同步更新后端文件。
 */

const KB = 1024
const MB = 1024 * 1024
const GB = 1024 * 1024 * 1024

export const VIDEO_LIMITS = {
  tencent_video: { minDuration: 5,    maxDuration: 5400,         maxSize: 20 * GB, maxTitleLength: 80 },
  iqiyi:         { minDuration: 5,    maxDuration: 3600,         maxSize: 16 * GB, maxTitleLength: Infinity },
  douyin:        { minDuration: 5,    maxDuration: 3600,         maxSize: 16 * GB, maxTitleLength: Infinity },
  baijiahao:     { minDuration: 5,    maxDuration: Infinity,     maxSize: 12 * GB, maxTitleLength: Infinity },
  weibo:         { minDuration: 0,    maxDuration: Infinity,     maxSize: 15 * GB, maxTitleLength: 30 },
  kuaishou:      { minDuration: 5,    maxDuration: 3600,         maxSize: 12 * GB, maxTitleLength: Infinity },
  bilibili:      { minDuration: 5,    maxDuration: 36000,        maxSize: 16 * GB, maxTitleLength: 80, maxDescLength: 2000 },
  xiaohongshu:   { minDuration: 5,    maxDuration: 14400,        maxSize: 20 * GB, maxTitleLength: 20 },
  channels:      { minDuration: 5,    maxDuration: 28800,        maxSize: 20 * GB, maxTitleLength: Infinity },
  tiktok:        { minDuration: 5,    maxDuration: 3600,         maxSize: 16 * GB, maxTitleLength: Infinity },
  youtube:       { minDuration: 5,    maxDuration: 36000,        maxSize: 16 * GB, maxTitleLength: Infinity },
  alipay:        { minDuration: 5,    maxDuration: Infinity,     maxSize: 8 * GB,  maxTitleLength: Infinity },   // 文档:≤8G,时长不限
  zhihu:         { minDuration: 0,    maxDuration: Infinity,     maxSize: Infinity, maxTitleLength: Infinity }, // 文档:时长大小不限
  // CSDN: 视频大小≤2G, 时长不限, 标题≤30字
  csdn:          { minDuration: 0,    maxDuration: Infinity,     maxSize: 2 * GB, maxTitleLength: 30 },
  // VIVO: 视频大小≤2G, 时长≤90min(5400s), 用「视频描述」非标题故不限标题
  vivo:          { minDuration: 0,    maxDuration: 5400,         maxSize: 2 * GB, maxTitleLength: Infinity },
  // 微信公众号: 视频时长<1h, 标题≤64字, 描述(含#标签)≤300字
  weixin_gzh:    { minDuration: 0,    maxDuration: 3600,         maxSize: Infinity, maxTitleLength: 64, maxDescLength: 300 },
  // 淘宝光合: 时长≤30min, 文件≤1.5G, 标题≤30字, 描述(含#标签)≤1000字
  taobao_guanghe: { minDuration: 0,   maxDuration: 1800,         maxSize: 1.5 * GB, maxTitleLength: 30, maxDescLength: 1000 },
  // 京东京麦: 标题 5~27 字
  jingmai:        { minDuration: 0,   maxDuration: Infinity,     maxSize: Infinity, minTitleLength: 5, maxTitleLength: 27 },
}

const PLATFORM_NAMES = {
  tencent_video: '腾讯视频',
  iqiyi: '爱奇艺',
  douyin: '抖音',
  baijiahao: '百家号',
  weibo: '微博',
  kuaishou: '快手',
  bilibili: 'B站',
  xiaohongshu: '小红书',
  channels: '视频号',
  tiktok: 'TikTok',
  youtube: 'YouTube',
  alipay: '支付宝',
  zhihu: '知乎',
  csdn: 'CSDN',
  vivo: 'VIVO',
  weixin_gzh: '微信公众号',
  taobao_guanghe: '淘宝光合',
  jingmai: '京东京麦',
}

export function formatSize(sizeBytes) {
  if (sizeBytes == null) return '-'
  if (!isFinite(sizeBytes) || sizeBytes < 0) return '未知'
  if (sizeBytes < KB) return `${sizeBytes.toFixed(1)} B`
  if (sizeBytes < MB) return `${(sizeBytes / KB).toFixed(1)} KB`
  if (sizeBytes < GB) return `${(sizeBytes / MB).toFixed(1)} MB`
  return `${(sizeBytes / GB).toFixed(1)} GB`
}

export function formatDuration(seconds) {
  if (seconds == null) return '-'
  if (!isFinite(seconds) || seconds < 0) return '未知'
  const s = Math.floor(seconds)
  if (s < 60) return `${s} 秒`
  const h = Math.floor(s / 3600)
  const m = Math.floor((s % 3600) / 60)
  const sec = s % 60
  if (h > 0) return `${h} 小时 ${m} 分 ${sec} 秒`
  return `${m} 分 ${sec} 秒`
}

function formatMaxDuration(max) {
  return max === Infinity ? '无限制' : formatDuration(max)
}

/**
 * 校验视频是否符合平台限制
 * @param {string} platformKey
 * @param {number} durationSec
 * @param {number} sizeBytes
 * @returns {{ ok: boolean, error: string }}
 */
export function validateVideoForPlatform(platformKey, durationSec, sizeBytes) {
  const limits = VIDEO_LIMITS[platformKey]
  if (!limits) return { ok: true, error: '' }

  const name = PLATFORM_NAMES[platformKey] || platformKey

  if (durationSec != null && durationSec < limits.minDuration) {
    return {
      ok: false,
      error: `${name}：时长 ${formatDuration(durationSec)} 小于最小值 (${formatDuration(limits.minDuration)})`,
    }
  }
  if (durationSec != null && durationSec > limits.maxDuration) {
    return {
      ok: false,
      error: `${name}：时长 ${formatDuration(durationSec)} 超出最大值 (${formatMaxDuration(limits.maxDuration)})`,
    }
  }
  if (sizeBytes != null && sizeBytes > limits.maxSize) {
    return {
      ok: false,
      error: `${name}：大小 ${formatSize(sizeBytes)} 超出限制 (最大 ${formatSize(limits.maxSize)})`,
    }
  }
  return { ok: true, error: '' }
}

/**
 * 按平台规则计算字符数：BMP 字符 = 1，emoji 等非 BMP 字符 = 3。
 * 用于标题/描述长度校验。JS 字符串的 .length 是 UTF-16 单元数（emoji 是 2），
 * 也不能直接用 codepoint 数（emoji 是 1），所以用遍历算。
 */
export function countCharsWithEmoji(s) {
  if (!s) return 0
  let n = 0
  for (const ch of s) {
    n += ch.codePointAt(0) > 0xFFFF ? 3 : 1
  }
  return n
}

/**
 * 校验标题是否符合平台限制
 * @param {string} platformKey
 * @param {string} title
 * @returns {{ ok: boolean, error: string, maxLength: number, actualLength: number }}
 */
export function validateTitleForPlatform(platformKey, title) {
  const limits = VIDEO_LIMITS[platformKey]
  if (!limits) return { ok: true, error: '', maxLength: Infinity, actualLength: 0 }
  const name = PLATFORM_NAMES[platformKey] || platformKey
  const min = limits.minTitleLength || 0
  const max = limits.maxTitleLength
  const len = countCharsWithEmoji(title)
  if (min && len < min) {
    return {
      ok: false,
      maxLength: max,
      actualLength: len,
      error: `${name}：标题至少 ${min} 个字 (当前 ${len} 字)`,
    }
  }
  if (max === Infinity) return { ok: true, error: '', maxLength: Infinity, actualLength: len }
  if (len > max) {
    return {
      ok: false,
      maxLength: max,
      actualLength: len,
      error: `${name}：标题 ${len} 字超过限制 (最多 ${max} 字,emoji 按 3 算)`,
    }
  }
  return { ok: true, error: '', maxLength: max, actualLength: len }
}

/**
 * 校验简介是否符合平台限制（如 B 站 ≤ 2000 字，emoji 按 3 算）
 * @param {string} platformKey
 * @param {string} desc
 * @returns {{ ok: boolean, error: string, maxLength: number, actualLength: number }}
 */
export function validateDescForPlatform(platformKey, desc) {
  const limits = VIDEO_LIMITS[platformKey]
  if (!limits || !limits.maxDescLength) return { ok: true, error: '', maxLength: Infinity, actualLength: 0 }
  const name = PLATFORM_NAMES[platformKey] || platformKey
  const max = limits.maxDescLength
  const len = countCharsWithEmoji(desc)
  if (max === Infinity) return { ok: true, error: '', maxLength: Infinity, actualLength: len }
  if (len > max) {
    return {
      ok: false,
      maxLength: max,
      actualLength: len,
      error: `${name}：简介 ${len} 字超过限制 (最多 ${max} 字,emoji 按 3 算)`,
    }
  }
  return { ok: true, error: '', maxLength: max, actualLength: len }
}
